src/Analyzers/MSTest.Analyzers/DoNotUseShadowingAnalyzer.cs

```diff
@@ -0,0 +1,113 @@
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+public sealed class DoNotUseShadowingAnalyzer : DiagnosticAnalyzer
+{
+    private static readonly LocalizableResourceString Title = new(nameof(Resources.DoNotUseShadowingTitle), Resources.ResourceManager, typeof(Resources));
+    private static readonly LocalizableResourceString Description = new(nameof(Resources.DoNotUseShadowingDescription), Resources.ResourceManager, typeof(Resources));
+    private static readonly LocalizableResourceString MessageFormat = new(nameof(Resources.DoNotUseShadowingMessageFormat), Resources.ResourceManager, typeof(Resources));
+
+    internal static readonly DiagnosticDescriptor DoNotUseShadowingRule = DiagnosticDescriptorHelper.Create(
+        DiagnosticIds.DoNotUseShadowingRuleId,
+        Title,
+        MessageFormat,
+        Description,
+        Category.Design,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+        = ImmutableArray.Create(DoNotUseShadowingRule);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(context =>
+        {
+            if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingTestClassAttribute, out INamedTypeSymbol? testClassAttributeSymbol))
+            {
+                context.RegisterSymbolAction(
+                    context => AnalyzeSymbol(context, testClassAttributeSymbol),
+                    SymbolKind.NamedType);
+            }
+        });
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol testClassAttributeSymbol)
+    {
+        var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+        if (namedTypeSymbol.TypeKind != TypeKind.Class
+            || !namedTypeSymbol.GetAttributes().Any(attr => SymbolEqualityComparer.Default.Equals(attr.AttributeClass, testClassAttributeSymbol)))
+        {
+            return;
+        }
+
+        INamedTypeSymbol? currentType = namedTypeSymbol.BaseType;
+        var baseClassesMembers = new List<ISymbol>();
+        while (currentType is not null)
+        {
+            baseClassesMembers.AddRange(currentType.GetMembers());
+            currentType = currentType.BaseType;
+        }
+
+        foreach (ISymbol member in namedTypeSymbol.GetMembers())
+        {
+            foreach (ISymbol baseMember in baseClassesMembers)
+            {
+                // Check if the member is shadowing a base class member
+                if (IsMemberShadowing(member, baseMember))
+                {
+                    context.ReportDiagnostic(member.CreateDiagnostic(DoNotUseShadowingRule, member.Name));
+                    return;
```

Comment:
You are only reporting one case at a time, so if there are multiple violations, the user has to fix, compile in chain. When possible we should show the various violations.